File: STRIVER/video/vid-5/pattern17/p17.py
#    A
#   ABA
#  ABCBA
# ABCDCBA

def p17(row:int)->None:

    for i in range(row):

        # space
        for j in range(row-i-1):
            print("-",end="")

        # stuff
        ch = ord('A')
        brk = ((2*i)+1) / 2
        for j in range(1, (2*i)+2):
            print(chr(ch),end="")
            if j <= brk:
                ch+=1
            else:
                ch-=1

        # space
        for j in range(row-i-1):
            print("-",end="")
        print(" ")

row = 4
p17(row)



# --------------------------------another type ---------------------------------------

# #    A   
# #   B B  
# #  C C C 
# # D D D D

# def p17(row:int)->None:

#     for i in range(row):

#         # space
#         for j in range(row-i-1):
#             print("-",end="")

#         # stuff
#         for j in range((2*i)+1):
#             ch = ord('A')+i
#             print(chr(ch),end="")

#         # space
#         for j in range(row-i-1):
#             print("-",end="")
#         print(" ")

# row = 4
# p17(row)

File: STRIVER/video/vid-5 (Patterns)/pattern17/p17.cpp
//    A
//   ABA
//  ABCBA
// ABCDCBA


#include <iostream>
using namespace std;

int main()
{
    int row = 4;

    for (char i = 0; i < row; i++){

        // space
        for(int j=0; j< row-i-1; j++){
            cout<<"-";
        }

        // stuff
        char ch = 'A';
        int brk = ((2*i)+1)/2;
        for(int j=1; j<= (2*i)+1; j++){
            cout<<ch;
            if(j<=brk) ch++;
            else ch--;
        }        

        // space
        for(int j=0; j< row-i-1; j++){
            cout<<"-";
        }

        cout<<endl;
    }

    return 0;
}